### PR TITLE
Add operator Smooth L1 Loss

### PR DESCRIPTION
**Pull request by precedenceguo**
Smooth L1 Loss is a loss first introduced in Fast R-CNN, which is a compromise of l1 and l2 losses. Parameter `sigma` is related to the turning point from square to linear. In default state `num_args` will be set as `2`, stating that only `data` and `target` are supplied to this loss function. For a more direct comparison with original caffe version, this implementation also supports two additional inputs:  `inside_weight` which scales input data and `outside_weight` which scales output loss.

### CODE REVIEW

**Comment by winstywang** [comment]
Since the numpy loss is too slow for training, we rewrite it in cpp to enable gpu. Please review and merge. @antinucleon @tqchen

**Comment by tqchen** [comment]
can you check if it is possible to use the new operator interface to simplify the loss?  e.g. https://github.com/dmlc/mxnet/blob/master/src/operator/loss_binary_op-inl.h

**Comment by precedenceguo** [comment]
On first glance I think it is possible. I will try to work on that soon.
In addition, I would say smooth l1 loss is similar to https://github.com/dmlc/mxnet/blob/master/src/operator/regression_output-inl.h. Is there a plan for the port to new interface of regression output?

**Comment by tqchen** [comment]
The old code will be kept for now. But we would encourage to move to the new one when possible as it is shorter to implement, and be able to expose to both ndarray and symbolic interface

**Comment by piiswrong** [comment]
@tqchen Could you update the create new operator document?

**Comment by precedenceguo** [comment]
We can reduce to 2 inputs and use element-wise multiplication to preserve the merits of inputs `inside_weight` and `outside_weight`. I had no luck finding out how to add a scalar input or a parameter beside two symbol inputs. In detail, I would like to know how to add a float parameter `sigma` to this operator. Could you please point me where to find that out? @tqchen

**Comment by precedenceguo** [comment]
Another problem lies in the composition of loss function.
```python
data = mx.sym.Variable('data')
inside_weight = mx.sym.Variable('inside_weight')
outside_weight = mx.sym.Variable('outside_weight')
sym = outside_weight * mx.sym.SmoothL1(name='bbox_loss', data=inside_weight * data)
```
Using ```executor.backward()``` would prompt ```Because the last operator is not Loss function, head_gradient is required in calling backward.``` Is there a way to resolve this problem?

**Comment by piiswrong** [comment]
Remove the dependency on out_grad in DeclareBackwardDependency will make the layer a loss layer

**Comment by tqchen** [comment]
If we want to make the behavior of the objective a true objective function. What we can do is

- Make sure the forward outputs a ```(1,)``` scalar of the loss function value
- Backprop as what we did in softmax entropy loss.

Then you can do composition like

```python
loss = alpha * softmax_cross_entroy(out1, label1) + beta * softmax_cross_entropy(out2, label2)
exec = loss.simple_bind()
head_grad = mx.nd.ones((1,), dev)
# propagate 1 as head gradient back, 
exec.backward([head_grad])
```

In your case, i guess you can likely do 
```python
data = mx.sym.Variable('data')
inside_weight = mx.sym.Variable('inside_weight')
outside_weight = mx.sym.Variable('outside_weight')
sym = outside_weight * mx.sym.SmoothL1(name='bbox_loss', data=inside_weight * data)
loss = mx.sym.sum(sym)

# propagate 1 as head gradient back, 
head_grad = mx.nd.ones((1,), dev)
exec.backward([head_grad])
```

**Comment by precedenceguo** [comment]
Thanks for comment. A dummy head gradient would make a composition of loss function work, thus making the utilization of new operator api possible. However, I wonder if this would compromise the top level training api, especially the distributed training feature.

**Comment by tqchen** [comment]
This won't compromise anything as long as we add support of head grad in
training API. So they would work for distributed setting. We can also add a
special operator that propagates 1 in backward pass to mark terminal loss

**Comment by winstywang** [comment]
@tqchen Does it mean we still need to pass gradient by ourselves? This is counter intuitive, and not friendly to users.

**Comment by winstywang** [comment]
Let's discuss a workaround here. Stay tuned.

**Comment by piiswrong** [comment]
@winstywang @precedenceguo This can be easily fixed with a MakeLoss layer that takes data_shape=(batch_size, 1) and outputs (1,) as summed loss. For backward, simply set in_grad to mx.nd.ones((batch_size, 1))

**Comment by precedenceguo** [comment]
That's very good to know. How to use `EnvArguments` to pass an extra parameter `float sigma`?

**Comment by tqchen** [comment]
@precedenceguo  You need to add set_enable_kwargs to true in registration.Then the EnvArguments will contain the kwargs, which you can use dmlc's parameter structure to initialize

**Comment by precedenceguo** [comment]
Came into template deduction problem using newer operator interface.
```cpp
mshadow::Tensor<xpu, 2, DType> mdata = data.get<xpu, 2, DType>(s);
mshadow::Tensor<xpu, 2, DType> mtarget = target.get<xpu, 2, DType>(s);
mshadow::Tensor<xpu, 2, DType> diff = mdata - mtarget;
```
First error is the minus operation. How to fix this?

**Comment by tqchen** [comment]
```
mshadow::Tensor<xpu, 2, DType> diff = mdata - mtarget;
```
diff need to be first get from some container, before you assign it

**Comment by precedenceguo** [comment]
Thanks for comment. I overlooked this and ```ScalarExp``` previously. I feel sorry to say that I didn't figure out how to access kwargs or initialize parameters with dmlc parameter structure from ```operator_utils.h```.  In this example, parameter ```DType sigma2 = 1.0``` should be replaced by the square kwargs ```sigma```. Could you please provide an example on this?

**Comment by piiswrong** [comment]
@precedenceguo Look into the operators. Do something like this:
```
struct ConcatParam : public dmlc::Parameter<ConcatParam> {
  int num_args;
  int dim;
  DMLC_DECLARE_PARAMETER(ConcatParam) {
    DMLC_DECLARE_FIELD(num_args).set_lower_bound(1)
    .describe("Number of inputs to be concated.");
    DMLC_DECLARE_FIELD(dim).set_range(0,  4).set_default(1)
    .describe("the dimension to be concated.");
  }
};  // struct ConcatParam
DMLC_REGISTER_PARAMETER(ConcatParam);
ConcatParam param_;
param_.Init(kwargs);
std::cout << param_.dim;
```

**Comment by precedenceguo** [comment]
How to make gradient of composition loss function work?
For ```inside_weight```, gradient is not right since the product ```inside_weight * (data - target)``` functions as input together. This can be fixed by making ```inside_weight``` an operand.
However,  there are problems with ```outside_weight```.
```python
data = mx.sym.Variable('data')
target = mx.sym.Variable('target')
inside_weight = mx.sym.Variable('inside_weight')
outside_weight = mx.sym.Variable('outside_weight')
sym = outside_weight * mx.sym.smooth_l1(name='loss',
        data=inside_weight * (data - target), scalar=1.0)
a = mx.nd.array([[1., 2., 3., 4., 5., 6.]])
b = mx.nd.array([[0.99, 0.5, 3.5, 2., 7., 10.]])
c = mx.nd.ones(a.shape)
d = mx.nd.array([[2, 2, 2, 2, 2, 2]])
executor = sym.simple_bind(ctx=mx.cpu(), data=a.shape, target=b.shape,
        inside_weight=c.shape, outside_weight=d.shape)
executor.forward(is_train=True, data=a, target=b, inside_weight=c, outside_weight=d)
head_grad = mx.nd.ones(executor.outputs[0].shape)
executor.backward(out_grads=[head_grad])
```
The output is 
```[[ 0.00999999  1.         -0.5         1.         -1.         -1.        ]]```
instead of 
```[[ 0.01999998  2.         -1.          2.         -2.         -2.        ]]```
i.e. missing the ```outside_weight```.

**Comment by piiswrong** [comment]
you don't need inside weight. Modify the backward of smoothl1 like this: ASSIGN_DISPATCH(igrad, req,
 +                    out_grad*F<mshadow_op::smooth_l1_gradient>(src, ScalarExp<DType>(sigma2)))

**Comment by precedenceguo** [comment]
Thanks for pointing that out. Numerical test stands sound now.

**Comment by piiswrong** [comment]
any updates on this? @precedenceguo

**Comment by precedenceguo** [comment]
Here is a summary of this pull request. Starting as a normal operator with variable argument number, a smooth l1 loss operator is presented. Then it is reformatted as a binary operator, utilizing the uniform operator framework. After that it is rewritten as an unary operator since it functions like a `sin`, or any other element-wise operator with same input and output shape.
We can composite what we have now to be use this as a loss in object detection
```python
sym = outside_weight * mx.sym.smooth_l1(data=inside_weight * (data - target), scalar=1.0)
sym = mx.sym.ElementWiseSum(sym, name='bbox_loss')
```
The second line is used to give a name to the composition loss. I have tested this composition loss and it gives the same output as the corresponding numpy loss.
I am not sure why the r-test in travis would fail.
I am not sure about why th

**Comment by piiswrong** [comment]
Thanks, great work.
One note: ElementWiseSum still needs head_grad. Are you feeding ones to backward?

@thirdwing Do you know why the r test is failing?

**Comment by thirdwing** [comment]
I know and it has been fixed in https://github.com/dmlc/mxnet/pull/2014/files yesterday.

**Comment by piiswrong** [comment]
@precedenceguo You just need to merge in the most recent code from master and travis test should be fixed

**Comment by precedenceguo** [comment]
Yes, a gradient of ones is feed to backward to make this composition work.
```python
head_grad = mx.nd.ones(executor.outputs[0].shape)
executor.backward([head_grad])
```

**Comment by piiswrong** [comment]
You can add a op that does nothing on foward and set in grad to 1 on
backward so that it can work with original api

**Comment by wp8619** [comment]
Then can this new op be used directly as a loss layer? In fast rcnn, the smoothl1loss needs to be grouped with cls loss, then how to use the "head_grad"?

**Comment by precedenceguo** [comment]
`out_grads` can be a list of NDArray for the case of multiple outputs. See `executor.backwards` declaration in `python/mxnet/executor.py`. What we are proposing here is a `MakeLoss` layer that does nothing but propagate ones of proper shape to every inputs. This special layer has no backward dependency and can be used as the last layer.

**Comment by wp8619** [comment]
@precedenceguo Thanks for your comment. Then the last layer can be defined as:
sym = outside_weight * mx.sym.smooth_l1(data=inside_weight * (data - target), scalar=1.0)
sym = mx.sym.ElementWiseSum(sym, name='bbox_loss')

with sym ready to be grouped with other symbol. Is it right?

**Comment by precedenceguo** [comment]
That's right. @wp8619

**Comment by wp8619** [comment]
Thanks very much. @precedenceguo

**Comment by wp8619** [comment]
Could you please give an example of how to write the MakeLoss layer? @precedenceguo

**Comment by precedenceguo** [comment]
Head gradient should have the same shape as output. The new `MakeLoss` layer can spare this effort and add a `grad_scale` to many unary and binary operators.
```python
loss = outside_weight * mx.sym.smooth_l1(data=inside_weight * (data - target), scalar=1.0)
bbox_loss = mx.sym.MakeLoss(name='bbox_loss', data=loss, grad_scale=1.0)
```

File: src/operator/make_loss-inl.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1) << "MakeLoss can only be used to one input";
+    CHECK_EQ(out_data.size(), 1);
+    CHECK_EQ(in_data[make_loss_enum::kData].ndim(), 2)
+    << "MakeLoss applies to all unary and binary operator with 2 dimension input";
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2> data = in_data[make_loss_enum::kData].get<xpu, 2, real_t>(s);
+    Tensor<xpu, 2> out = out_data[make_loss_enum::kOut].get<xpu, 2, real_t>(s);
+    Assign(out, req[make_loss_enum::kData], F<mshadow_op::identity>(data));
```
Comment (piiswrong) [inline src/operator/make_loss-inl.h]: You can drop the F<mshadow_op::identity>.
You can even save a copy if req[.] is WriteInplace which should be the case most of the times.
It's not a big deal though.

**Comment by piiswrong** [comment]
LGTM now.
@precedenceguo Is this ready to be merged?

**Comment by precedenceguo** [comment]
Yes, I think so. Please review and consider.

**Comment by MissFishLi** [comment]
hi, thanks for the commit!
I wonder which kind of evaluating metrics should I use when using make loss op?

**Comment by precedenceguo** [comment]
It depends on what to evaluate. For smooth l1 loss, there is no need for special metric since raw output is enough (the loss value).

**Comment by piiswrong** [comment]
We should have a DummyMetric that just print the output value. The Torch
metric sort of does this. Can can make a alias and better document it

**Comment by MissFishLi** [comment]
Thanks a lot! @precedenceguo

**Comment by MissFishLi** [comment]
Sorry I don't exactly understand how this works yet. 
If the smooth l1 op produced a (w, h) tensor, and make loss op take it as an input. So, the output tensor will also be (w, h) dim. Do you mean that this output tensor is the loss value?  And if i want to get the total loss value(a single number), should I sum the output tensor? @precedenceguo

**Comment by precedenceguo** [comment]
The output is element-wise loss. Sum the output will give a single number.
